feed.xml

```diff
@@ -14,6 +14,436 @@ Nick Ripley's blog. Programming and other interests.
+<ul >
+<li >
+First, read the <a href="https://go.dev/doc/asm">Go assembly guide</a>.
+</li>
+<li >
+Prefer an assembly generator like <a href="https://github.com/mmcloughlin/avo" target="_blank">Avo</a>.
+Tools like Avo are aware of the underlying rules for using registers and manipulating the stack,
+and generally make writing non-trivial amounts of assembly easier.
+</li>
+<li >
+If your assembly function calls Go functions,
+prefer not to use the frame pointer registers at all
+(<code >BP</code> for AMD64, <code >R29</code> for ARM64).
+Leaving a non-frame pointer value in those registers prior to calling a Go function can crash the runtime execution tracer and profilers.
+</li>
+<li >
+Otherwise, if your function has a non-zero frame size,
+the assembler will correctly make a stack frame for you,
+growing the stack if needed,
+and properly save and restore the frame pointer.
+The means the frame pointer will be safe to use.
+Consider giving your function a small, non-zero frame size even if you don't <i >need</i> to if you want to use the frame pointer.
+<sup ><a href="https://nsrip.com/posts/clobberfp.html#footnote1" id="back1">[1]</a></sup>
+</li>
+<li >
+Otherwise, if you need to use the frame pointer and don't want to have a frame,
+save the original frame pointer on the stack yourself and restore it when the function returns.
+Take care on ARM64:
+Go compiled functions will save the frame pointer one word below their stack frame,
+so the first 8 bytes of stack frame are off-limits,
+and the stack pointer must be 16-byte aligned.
+</li>
+</ul>
+</p>
+
+
+<div class="hover">
+<h2 id="register">Clobbering a register <a href="https://nsrip.com/posts/clobberfp.html#register" class="hide">&sect;</a></h2>
+</div>
+
+<p >
+The first issue was reported at <a href="https://github.com/golang/go/issues/69629" target="_blank">go.dev/issue/69629</a>.
+After upgrading to Go 1.23, <i >and</i> building with <a href="https://go.dev/doc/pgo" target="_blank">profile-guided optimization</a> (PGO),
+the block profiler consistently crashed when the program was under load.
+The crash happened when collecting a call stack via frame pointer unwinding.
+This affected the program when built for the <code >amd64</code> architecture,
+but not when it was built for <code >arm64</code>.
+I had <a href="https://go.dev/cl/533258" target="_blank">contributed</a> frame pointer unwinding for the block and mutex profilers for Go 1.23,
+so I took a look at the issue.
+</p>
+
+<p >
+I won't go into the full investigation here.
+There were many dead ends.
+Thanks to the issue reporter, Liz Fong-Jones (@lizthegrey), for being so willing to collect and share information.
+The key clue turned out to be the address of the faulting memory access, <code >0x30bc5b31</code>,
+which was the same in every occurrence of the crash.
+The program was reading 8 bytes above a frame pointer,
```

Comment:
NIT: Maybe link "8 bytes above a frame pointer" to https://github.com/golang/go/blob/4b652e9f5f5c0793f2e41cd2876bce5a241b2c95/src/runtime/mprof.go#L592

Reply:
Good call, I'll push an update